recovery/src/main/java/com/kin/ecosystem/recovery/backup/presenter/BackupPresenterImpl.java

```diff
@@ -1,37 +1,66 @@
 package com.kin.ecosystem.recovery.backup.presenter;
 
 
+import android.os.Bundle;
+import android.support.annotation.Nullable;
 import com.kin.ecosystem.recovery.backup.view.BackupView;
 import com.kin.ecosystem.recovery.base.BasePresenterImpl;
+import com.kin.ecosystem.recovery.events.CallbackManager;
 
 public class BackupPresenterImpl extends BasePresenterImpl<BackupView> implements BackupPresenter {
 
+	private @Step
+	int step = STEP_START;
+	private final CallbackManager callbackManager;
+	private boolean isBackupSucceed = false;
+
+
+	public BackupPresenterImpl(CallbackManager callbackManager) {
+		this.callbackManager = callbackManager;
+	}
 
 	@Override
 	public void onAttach(BackupView view) {
 		super.onAttach(view);
-		setStep(STEP_START);
+		setStep(step, null);
 	}
 
 	@Override
 	public void onBackClicked() {
-		if (view != null) {
-			view.backButtonClicked();
+		if (step == STEP_WELL_DONE) {
+			setStep(STEP_CLOSE, null);
+		} else {
+			if (view != null) {
+				if(!isBackupSucceed && step == STEP_CREATE_PASSWORD) {
+					callbackManager.sendCancelledResult();
+				}
+				step--;
+				view.onBackButtonClicked();
+			}
 		}
 	}
 
 	@Override
-	public void setStep(@Step final int step) {
+	public void setStep(@Step final int step, @Nullable Bundle data) {
```

Comment:
instead of calling setStep with null you can add public void setStep(@Step final int step) that will call the second one with null